### PR TITLE
New: spaced-comment rule (fixes #1088)

### LINKED ISSUE #1088: Rule: Space after line comments
- name: __line-comments-spaces__
- params: \[0, "always"|"never"] ("always" by default)

good:

```javascript
var foo; // abc
var bar; /* a */

var foo(a/* height */) {
}
```

bad:

```javascript
var foo; //abc
var bar; /*a*/

var foo(a/*height*/) {
}
```

Don't know if inverse option needed or not.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/3104924-rule-space-after-line-comments?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by gyandeeps**

### CODE REVIEW

File: docs/rules/spaced-comment.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+This rule will enforce consistency of spacing after the start of a comment `//` or `/*`.
+
+This rule takes two arguments. If the first is `"always"` then the `//` or `/*` must be followed by at least once whitespace.
+If `"never"` then there should be no whitespace following.
+The default is `"always"`.
+
+The second argument is an object with one key, `"exceptions"`.
+The value is an array of string patterns which are considered exceptions to the rule.
+It is important to note that the exceptions are ignored if the first argument is `"never"`.
+Exceptions cannot be mixed.
```
Comment (ilyavolodin) [inline docs/rules/spaced-comment.md]: I think this needs to be rephrased.

File: lib/rules/spaced-comment.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    // Grab the exceptions array and build a RegExp matcher for it
+    var hasExceptions = context.options.length === 2;
+    var unescapedExceptions = hasExceptions ? context.options[1].exceptions : [];
+    var exceptions;
+
+    if (unescapedExceptions.length) {
+        exceptions = unescapedExceptions.map(function(s) {
+            return s.replace(/([.*+?${}()|\^\[\]\/\\])/g, "\\$1");
+        });
+        exceptionMatcher = new RegExp("(^(" + exceptions.join(")+$)|(^(") + ")+$)");
```
Comment (ilyavolodin) [inline lib/rules/spaced-comment.js]: If I'm reading this regex right, this only allows repeated exceptions of one type, yet in example I see mixed `+` and `-`. If the idea that you can use any exceptions in a single line without space then why not use
```js
/^[" + exceptions.join() + "]+$/
```

Comment (gyandeeps) [inline lib/rules/spaced-comment.js]: I am rookie when it comes to regex. I have borrowed most of the code form `spaced-line-comment` rule.
So I am not sure what difference your proposal would make. I can try.

File: lib/rules/spaced-comment.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+        if (requireSpace) {
+
+            // If length is zero, ignore it
+            if (node.value.length === 0) {
+                return;
+            }
+
+            // Space expected and not found
+            if (node.value.indexOf(" ") !== 0 && node.value.indexOf("\t") !== 0) {
```
Comment (ilyavolodin) [inline lib/rules/spaced-comment.js]: This isn't going to be able handle the following case (> is linebreak, _ is space)
```js
/*>
 *_Test>
 */
```

File: lib/rules/spaced-comment.js (hunk 2 above)
Comment (ilyavolodin) [inline lib/rules/spaced-comment.js]: My regex is going to allow only the characters in the exceptions list in a given string. If I'm reading your regex correctly, it only allows strings of one exception character repeated one or more times. If that was the intention, then my regex isn't going to work, but examples and unit tests show different exception characters used on the same line.

Comment (ilyavolodin) [inline lib/rules/spaced-comment.js]: Nevermind. I read examples wrong. It only allows one exception of a type per line. Not sure why this limitation exists, but that's fine.

Comment (gcochard) [inline lib/rules/spaced-comment.js]: It was an [explicit decision](https://github.com/eslint/eslint/issues/1345#issuecomment-58126556) when I originally wrote the rule.

File: tests/lib/rules/spaced-comment.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            options: ["always"]
+        },
+        {
+            code: "var a = 1; /*######*/",
+            options: ["always", {
+                exceptions: ["-", "=", "*", "#", "!@#"]
+            }]
+        },
+        {
+            code: "var a = 1; /*A valid comment starting with space */",
```
Comment (gcochard) [inline tests/lib/rules/spaced-comment.js]: This is misleading, should say "not starting with a space".

File: tests/lib/rules/spaced-comment.js [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+            options: ["always", {
+                exceptions: ["-", "=", "*", "!@#"]
+            }],
+            errors: [{
+                message: "Expected exception block, space or tab after /* in comment.",
+                type: "Block"
+            }]
+        },
+        {
+            code: "var a = 1; /*A valid comment starting with space */",
```
Comment (gcochard) [inline tests/lib/rules/spaced-comment.js]: Also needs "not starting..."

File: tests/lib/rules/spaced-comment.js (hunk 4 above)
Comment (gyandeeps) [inline tests/lib/rules/spaced-comment.js]: will make that change as i din't pay attention to what the comment was.

File: lib/rules/spaced-comment.js (hunk 2 above)
Comment (nzakas) [inline lib/rules/spaced-comment.js]: Yeah, let's focus on porting the current behavior here. We can file other issues for enhancements later.

**Pull request merged**